### PR TITLE
Add new EF Core data model (v2)

### PR DESCRIPTION
**Pull request by EraYaN**
Do-over of #1431 (click for old discussion)

**Changes**
For now it adds a project with an [Entity Framework Visual Editor](https://marketplace.visualstudio.com/items?itemName=michaelsawczyn.EFDesigner) based entity model. Note this only contains some of the media types and some of the fields, those are easy to add later.

My main philosophy for this is to keep the database structure as restrictive as possible, as this forces good "data-hygiene" as it were. So no JSON blobs, no generic tables that are key-value stores (if we want that we should use Redis or something) etc.

**Images History**
<details>
<summary>Old Revisions</summary>
2019-06-01 (b82a749)
diagram_2019-06-01_001

2019-06-01 (17f4034)
diagram_2019-06-01_002

2019-07-06 (8e7d876)
diagram_2019-07-06_001

2019-07-07 (b59cd27)
diagram_2019-07-07_001

2019-09-04 (16c4058)
diagram_2019-09-04_003

2019-09-14 (95658e2)
diagram_2019-09-14_001
</details>

2020-04-26 (03c8344)
diagram_2020-04-26_001

TODO LIST:
- [ ]  #2517 - Add user related permissions and preference from existing database schema (so far, only xml based ones have been added)
- [ ]  #2518 - Add user linked tracking of watch status and resume points for media in a media agnostic way (aka, allow for resuming audiobooks and comics as well as movies and tv shows)
- [ ] #2519 - Implement storage of system wide settings. Avoid "node" specific settings like bound IP, ports, hw accel, etc while migrating settings like default UI culture, metadata language, etc
- [ ] #2520 - Implement storage of multiple languages for metadata (including images). This would allow admins to define server wide fetch languages allowing for better internationalization support. The admin could say they want 3 languages and then users can decide which order they want these languages to be when displaying metadata.
- [ ] #2521 - Implement user definable fallback that can have multiple steps for UI and metadata language. Currently, US English is the only fallback if you chose a language other than it. Ideally, users should be able to fall back through a list of languages of their choosing (with US English being the final fallback as it is now)
- [ ] #2522 - Implement a generic grouping and collections system that will allow users and admins to group media in arbitrary ways. As an example, you could group Futurama into 3 parts, pre-movies, movies, and post movies then group these together such that if you watch the last episode of pre-movies "up next" appears shows the first movie. But... Ordering must not be required. Maybe a user just wants to group the 300 Scooby-Doo movies into one "folder." This should effectively take over the role of playlists since it serves the same purpose but is more flexible

### CODE REVIEW

**Comment by sparky8251** [comment]
@uffcd Since the other issue was accidentally borked and this new one opened, ill answer your question here.

> Would it be possible to track watched status for item+library? So that I dont get multiple entries for everything watched because there is the same movie in several libraries (f.ex. in Movies and Movies4K and Movies french)

This is how it will likely work, since we will be dramatically improving the usefulness of collections as part of this schema. If you look, the item and its metadata already exists largely independent of the library itself (metadata has some links to library). My hope is that this schema is flexible rather than rigid, so we can continue to improve JF for the foreseeable future so of course I am going to try what you have asked for :)

**Comment by Artiume** [comment]
We should write a guide on how to get this up and running in VS so it's easier for new people to contribute :)

**Comment by Ornias1993** [comment]
@Artiume 
I think the target of this PR is primarily people with somewhat of a background in relational databases (thats also what those schema's are aimed at). People with the knowhow to understand these images are perfectly able to collaborate I expect.

**Comment by Artiume** [comment]
I did `git clone --single-branch --branch new-data-model https://github.com/EraYaN/jellyfin.git` to clone the database and then opened it using Visual Studios. It is currently installing packages for VS. 

@nielsvanvelzen let me know if these are sufficient instructions.

**Comment by sparky8251** [comment]
So we have some notes to reference later on when the rewrite comes, I will place comments here that contain a deep dives on odd settings as I look to shore up the final parts of the schema rewrite.

First up is in the `user.db`, the items attached to each user in the JSON blob. 

* `ImageInfos` Looking into this, it seems all images share the same unified API and thusly contains information that is irrelevant to the user API like height, width, allowing for multiple images to be returned, etc. We should replace this with more specific endpoints so they can be more accurately tailored to each specific instance.

* `IsHD` seems to be attached to *everything* in the database including the user because its part of the `BaseItem`. It seems to allow you to sort media and TV channels by HD/Non-HD and has strange criteria for determining if something is HD or not. All non-Live TV media seems to assume anything with a height more than 720 `IsHD` even though this isn't necessarily true depending on the media, and Live TV seems to get this info from Schedules Direct. I am unsure if this is API end point is used by any clients currently, but it should be reworked to no longer be an option for music and books and allow for more customization of detection of HD media.

**Comment by Sp4rky001** [comment]
> @uffcd Since the other issue was accidentally borked and this new one opened, ill answer your question here.
> 
> > Would it be possible to track watched status for item+library? So that I dont get multiple entries for everything watched because there is the same movie in several libraries (f.ex. in Movies and Movies4K and Movies french)
> 
> This is how it will likely work, since we will be dramatically improving the usefulness of collections as part of this schema. If you look, the item and its metadata already exists largely independent of the library itself (metadata has some links to library). My hope is that this schema is flexible rather than rigid, so we can continue to improve JF for the foreseeable future so of course I am going to try what you have asked for :)
> 
> The hardest part will be supporting of turning this off, so those that want them treated separately can do so...

Perhaps 4K, 1080p, 720p, etc should simply be considered the "quality" of the stream.  Just as on youtube, you can choose your video quality.  The reason is because I think JF needs to look to a future function of allowing encoding movies/shows into lower "mobile-friendly" formats ahead of time as opposed to on-the-fly transcoding, just as Plex now has the "optimize" option.

Having the ability to pre-encode at the expense of additional disk space will mean the server doesn't need to transcode the same movie or show each time it is watched, saving on CPU/GPU cycles (and the number of disk read/writes in the case of an SSD).  Perhaps a predictive algorithm can even be developed to pre-encode according to "up next" or "continue watching", and a cleanup algorithm that can delete those encodes once users have not watched a movie or show for a period of time.

**Comment by JustAMan** [comment]
> Perhaps 4K, 1080p, 720p, etc should simply be considered the "quality" of the stream. Just as on youtube, you can choose your video quality. The reason is because I think JF needs to look to a future function of allowing encoding movies/shows into lower "mobile-friendly" formats ahead of time as opposed to on-the-fly transcoding, just as Plex now has the "optimize" option.

I don't think we should be having this in the core server, but it would make an awesome plugin.
Not sure it's currently possible to implement it though, as I haven't looked into plugin system much.

**Comment by cromefire** [comment]
Also how is the "highest quality" calculated? I've often 3 Versions, all 1080p, but one in h264, hevc and vp9 with vp9 being the highest quality.

**Comment by dkanada** [comment]
Here are some thoughts about the existing schema.

- DateAdded and DateModified should also be tracked on the files
- MediaFileKind what is the Sidecar type
- MediaFileKind AlternativeFormat -> AdditionalFormat
- Artwork -> Image (artwork isn't technically accurate for some of the media types)
- ImageType -> Chapter (one example for why artwork should be image instead)
- Good idea making the collections a linked list
- MediaFileStream -> MediaStream so track can inherit Audio for the ffmpeg information

**Comment by EraYaN** [comment]
Alright I have cleaned up all binary and other model files. This is the "pure" C# version, so everyone can work on it. And we should probably merge it to get going.

**Comment by Ornias1993** [comment]
You're on fire! :)

**Comment by mark-monteiro** [comment]
To clarify, all this PR contains is generated code? From a quick glance the generated code is all in a new project and isn't used anywhere. And the `.efmodel` file used by the Visual Editor extension to generate the code doesn't appear to be included, so the plan is to do manual code changes to the entity class files from here on in?

**Comment by EraYaN** [comment]
Yes manual changes from here on out. The efmodel files switched to binary in a new release, making git absolutely useless with them. And frankly there is enough boiler plate now to see/copy how to add new fields.

**Comment by mark-monteiro** [comment]
Is it possible to customize how the code is generated? I'm trying to decide if I should be doing a code review of the generated classes to see if we want to alter them in any way. Probably easiest to make a change like that now while we haven't broken away from the efmodel file

**Comment by barronpm** [comment]
I plan on making a PR to move everything to `Jellyfin.Model`, as well as document the code and do any manual tweaking to make the code more readable. I think that any changes should be made in that PR

**Comment by EraYaN** [comment]
There is no easy way to control the code generation. This should just be the first entry of the code (that way we can fully track changes).

**Comment by mark-monteiro** [comment]
> There is no easy way to control the code generation. This should just be the first entry of the code (that way we can fully track changes).

It looks like there is a path to doing this that doesn't seem too bad: https://msawczyn.github.io/EFDesigner/Customizing.html#replacing-the-t4-templates. Have we tried this yet? It seems like a much more scalable solution that making manual changes to ~50 files. We could merge this PR with the T4 templates added to the repository, and then remove them once the code generation is at a place we like.

**Comment by EraYaN** [comment]
I mean I know, but the T4 templates were not in the repo the only template used subtemplates from the extension. Easy is not a word I'd describe it. Changes can be made when we implement that explicit object, but if we want the T4's in this repo we might need to extract them from the extension.

File: MediaBrowser.sln [hunk 1]
```diff
@@ -176,24 +178,17 @@ Global
...
-		UpdateAssemblyVersion = True
-		UpdateAssemblyFileVersion = True
-		UpdateAssemblyInfoVersion = True
-		AssemblyVersionSettings = None.None.None.None
-		AssemblyFileVersionSettings = None.None.None.None
-		AssemblyInfoVersionSettings = None.None.None.None
-		UpdatePackageVersion = False
-		AssemblyInfoVersionType = SettingsVersion
-		InheritWinAppVersionFrom = None
-	EndGlobalSection
```
Comment (mark-monteiro) [inline MediaBrowser.sln:196]: What's with these changes here?

**Comment by barronpm** [comment]
In regards to migrations, they won't be needed until we make further changes to the schema. That being said, I plan on adding the initial migration code with the user database EF Core PR.

**Comment by mark-monteiro** [comment]
> I plan on making a PR to move everything to `Jellyfin.Model`, as well as document the code and do any manual tweaking to make the code more readable. I think that any changes should be made in that PR

Cool, makes sense! Please be sure to split this up onto multiple PRs though. Maybe one to move everything to `Jellyfin.Model`, a second with some manual tweaking on just one of the model files so that we can code review it quickly, and then a third applying the tweaks to the rest of the model files.

File: MediaBrowser.sln (hunk 1 above)
Comment (EraYaN) [reply MediaBrowser.sln:196]: Right those shouldn't be there.

**Pull request merged**